TESTS/mbedmicro-rtos-mbed/signals/main.cpp

```diff
@@ -15,49 +15,357 @@
  */
 #include "mbed.h"
 #include "greentea-client/test_env.h"
+#include "unity.h"
+#include "utest.h"
 #include "rtos.h"
 
+using namespace utest::v1;
+
+
 #if defined(MBED_RTOS_SINGLE_THREAD)
   #error [NOT_SUPPORTED] test not supported
 #endif
 
-#define TEST_STACK_SIZE 512
+#define TEST_STACK_SIZE   512
+#define ALL_SIGNALS 0x7fffffff
+#define PROHIBITED_SIGNAL 0x80000000
+#define MAX_FLAG_POS 30
+
+Semaphore sem_a(0, 1);
+Semaphore sem_b(0, 1);
+volatile int32_t ret_val;
+volatile osStatus ret_status;
+
+
+template <int32_t signals, uint32_t timeout>
+void run_signal_wait(void)
+{
+    osEvent ev = Thread::signal_wait(signals, timeout);
+    ret_status = ev.status;
+    sem_a.release();
+}
+
+template <int32_t signals, uint32_t timeout>
+void run_signal_wait2(void)
+{
+    sem_a.release();
+    osEvent ev = Thread::signal_wait(signals, timeout);
+    ret_status = ev.status;
+    sem_a.release();
+}
+
+template <int32_t signals, uint32_t timeout>
+void run_signal_wait3(void)
+{
+    sem_a.release();
+    sem_b.wait();
+    osEvent ev = Thread::signal_wait(signals, timeout);
+    ret_status = ev.status;
+    sem_a.release();
+}
+
+template <int32_t signals>
+void run_clear(void)
+{
+    ret_val = Thread::signal_clr(signals);
+    sem_a.release();
+}
+
+template <int32_t signals>
+void run_wiat_clear(void)
+{
+    sem_a.release();
+    sem_b.wait();
+    ret_val = Thread::signal_clr(signals);
+    sem_a.release();
+}
+
+template <int32_t signals1, int32_t signals2>
+void run_double_wiat_clear(void)
+{
+    sem_a.release();
+    sem_b.wait();
+    ret_val = Thread::signal_clr(signals1);
+    sem_a.release();
+    sem_b.wait();
+    ret_val = Thread::signal_clr(signals2);
+    sem_a.release();
+}
+
+void run_loop_wiat_clear(void)
+{
+    for (int i = 0; i <= MAX_FLAG_POS; i++) {
+        sem_b.wait();
+        ret_val = Thread::signal_clr(0);
+        sem_a.release();
+    }
+}
+
+
+/** Test signal_clr with no signals specified
+
+    Given two threads A and B running, B with all signals already set
+    When execute @a signal_clr(0) with no signals specified
+    Then thread B @a signal_clr status should be ALL_SIGNALS indicating that thread B state is unchanged
+ */
+void test_clear_0(void)
+{
+    int32_t ret;
+    Thread tb(osPriorityNormal, TEST_STACK_SIZE);
+    tb.start(callback(run_double_wiat_clear<0, 0>));
+
+    sem_a.wait();
+    ret = tb.signal_set(ALL_SIGNALS);
+    TEST_ASSERT_EQUAL(ALL_SIGNALS, ret);
+
+    sem_b.release();
+    sem_a.wait();
+    TEST_ASSERT_EQUAL(ALL_SIGNALS, ret_val);
+
+    sem_b.release();
+    sem_a.wait();
+    TEST_ASSERT_EQUAL(ALL_SIGNALS, ret_val);
+}
+
+/** Test if any signals are set on just created thread
+
+    Given one thread running
```

Comment:
`is running` or `one running thread` -> and something similar for other descriptions